safekeeper/spec/ProposerAcceptorStatic.tla

```diff
@@ -0,0 +1,448 @@
+      \* flush_lsn in case of acceptor. In case of proposer it is the current
+      \* writing position, but it can't be less than flush_lsn, so we
+      \* take flush_lsn.
+      acc_common_term_end == IF last_common_idx = Len(acc_th) THEN acc_flush_lsn ELSE acc_th[last_common_idx + 1].lsn
+      prop_common_term_end == IF last_common_idx = Len(prop_th) THEN acc_flush_lsn ELSE prop_th[last_common_idx + 1].lsn
+    IN
+      [term |-> last_common_term, lsn |-> Min(acc_common_term_end, prop_common_term_end)]
+
+\* Elected proposer p immediately truncates WAL (and term history) of acceptor a
+\* before starting streaming. Establishes nextSendLsn for a.
+\*
+\* In impl this happens at each reconnection, here we also allow to do it multiple times.
+TruncateWal(p, a) ==
+  /\ prop_state[p].state = "leader"
+  /\ acc_state[a].term = prop_state[p].term
+  /\ LET
+       hcp == FindHighestCommonPoint(prop_state[p].termHistory, AcceptorTermHistory(a), FlushLsn(a))
+       next_send_lsn == (a :> hcp.lsn) @@ prop_state[p].nextSendLsn
+     IN
+          \* Acceptor persists full history immediately; reads adjust it to the
+          \* really existing wal with AcceptorTermHistory.
+       /\ acc_state' = [acc_state EXCEPT ![a].termHistory = prop_state[p].termHistory,
+                                         \* note: SubSeq is inclusive, hence -1.
+                                         ![a].wal = SubSeq(acc_state[a].wal, 1, hcp.lsn - 1)
+                       ]
+       /\ prop_state' = [prop_state EXCEPT ![p].nextSendLsn = next_send_lsn]
+       /\ UNCHANGED <<committed, elected_history>>
+
+\* Append new log entry to elected proposer
+NewEntry(p) ==
+  /\ prop_state[p].state = "leader"
+  /\ LET
+       \* entry consists only of term, index serves as LSN.
+       new_entry == prop_state[p].term
+     IN
+       /\ prop_state' = [prop_state EXCEPT ![p].wal = Append(prop_state[p].wal, new_entry)]
+       /\ UNCHANGED <<acc_state, committed, elected_history>>
+
+\* Immediately append next entry from elected proposer to acceptor a.
+AppendEntry(p, a) ==
+  /\ prop_state[p].state = "leader"
+  /\ acc_state[a].term = prop_state[p].term
+  /\ a \in DOMAIN prop_state[p].nextSendLsn \* did TruncateWal
+  /\ prop_state[p].nextSendLsn[a] <= Len(prop_state[p].wal) \* have smth to send
+  /\ LET
+       send_lsn == prop_state[p].nextSendLsn[a]
+       entry == prop_state[p].wal[send_lsn]
+       \* Since message delivery is instant we don't check that send_lsn follows
+       \* the last acc record, it must always be true.
+     IN
+       /\ prop_state' = [prop_state EXCEPT ![p].nextSendLsn[a] = send_lsn + 1]
+       /\ acc_state' = [acc_state EXCEPT ![a].wal = Append(acc_state[a].wal, entry)]
+       /\ UNCHANGED <<committed, elected_history>>
+
+\* LSN where elected proposer p starts writing its records.
+PropStartLsn(p) ==
+    prop_state[p].termHistory[Len(prop_state[p].termHistory)].lsn
```

Comment:
Can we `Assert` that proposer is in state `"leader"`?

Reply:
added, forgot to comment here